### PR TITLE
feat(topology): show vm node details in the sidepanel

### PR DESCRIPTION
**Pull request by its-mitesh-kumar**
**Resolves:**
https://issues.redhat.com/browse/RHIDP-2893

### CODE REVIEW

File: plugins/topology/src/components/Topology/TopologySideBar/TopologyVirtualMachineDetails.tsx [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+      >
+        <TopologySideBarDetailsItem
+          label="Description"
+          emptyText="Not available"
+        >
+          {description}
+        </TopologySideBarDetailsItem>
+        <TopologySideBarDetailsItem label="Operating System">
+          Need to add it
+        </TopologySideBarDetailsItem>
```
Comment (jerolimov) [inline plugins/topology/src/components/Topology/TopologySideBar/TopologyVirtualMachineDetails.tsx:65]: Can we add or remove this???

File: plugins/topology/src/components/Topology/TopologySideBar/TopologyVirtualMachineDetails.tsx [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+        <TopologySideBarDetailsItem label="Node" emptyText="Not available">
+          {nodeName}
+        </TopologySideBarDetailsItem>
+        <TopologySideBarDetailsItem label="Workload profile">
+          {vm.spec?.strategy?.type}
+        </TopologySideBarDetailsItem>
+        <TopologySideBarDetailsItem label="User credentials">
+          {vm.spec?.strategy?.type}
+        </TopologySideBarDetailsItem>
```
Comment (jerolimov) [inline plugins/topology/src/components/Topology/TopologySideBar/TopologyVirtualMachineDetails.tsx:93]: This are not the user credentials, please update or remove this.

File: plugins/topology/src/components/Topology/TopologySideBar/TopologyVirtualMachineDetails.tsx [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+        <TopologySideBarDetailsItem label="SSH access">
+          {/* <span data-test="details-item-ssh-access-port">
+          {vmiReady ? 
+            sshServicesRunning ?  `Port: ${sshServicePort}` 
+             : 'SSH service disabled'
+            
+          : 
+            'Virtual machine not running')
+          }
+        </span> */}
```
Comment (jerolimov) [inline plugins/topology/src/components/Topology/TopologySideBar/TopologyVirtualMachineDetails.tsx:103]: You will improve this?

File: plugins/topology/src/components/Topology/TopologySideBar/TopologyVirtualMachineDetails.tsx [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+        </TopologySideBarDetailsItem>
+        <TopologySideBarDetailsItem label="Hardware devices">
+          <div>GPU devices : {getGPUDevices(vm).length}</div>
+          <div>Host devices : {getHostDevices(vm).length}</div>
+        </TopologySideBarDetailsItem>
```
Comment (jerolimov) [inline plugins/topology/src/components/Topology/TopologySideBar/TopologyVirtualMachineDetails.tsx:108]: Please remove the spaces after the "keys"

```suggestion
        <TopologySideBarDetailsItem label="Hardware devices">
          <div>GPU devices: {getGPUDevices(vm).length}</div>
          <div>Host devices: {getHostDevices(vm).length}</div>
        </TopologySideBarDetailsItem>
```

File: plugins/topology/src/vm-models.ts [hunk 5]
```diff
@@ -5,10 +5,27 @@ export const VirtualMachineGVK: GroupVersionKind = {
...
+
+export const VirtualMachineInstanceGVK: GroupVersionKind = {
+  apiVersion: 'v1',
+  apiGroup: 'kubevirt.io',
+  kind: 'VirtualMachineInstance',
+};
+// export const ReplicationControllerGVK: GroupVersionKind = {
+//   apiVersion: 'v1',
+//   kind: 'ReplicationController',
+// };
```
Comment (jerolimov) [inline plugins/topology/src/vm-models.ts:17]: We don't show RepilicaControllers at the moment?

File: plugins/topology/src/vm-models.ts [hunk 6]
```diff
@@ -5,10 +5,27 @@ export const VirtualMachineGVK: GroupVersionKind = {
...
 export const VirtualMachineModel: Model = {
   ...VirtualMachineGVK,
   abbr: 'VM',
   labelPlural: 'VirtualMachines',
   color: '#2b9af3',
   plural: 'virtualmachines',
 };
+export const VirtualMachineInstanceModel: Model = {
```
Comment (jerolimov) [inline plugins/topology/src/vm-models.ts:25]: Please add empty spaces between exported resources.

**Review by jerolimov — CHANGES_REQUESTED** [review]
Some notes also if this is in draft status:

File: plugins/topology/src/components/Topology/TopologySideBar/TopologyVirtualMachineDetails.tsx [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+      <div className="topology-workload-details">
+        <TopologyWorkloadDetails resource={vm}>
+          <TopologySideBarDetailsItem label="Status">
+            {vm?.status?.printableStatus}
+          </TopologySideBarDetailsItem>
+        </TopologyWorkloadDetails>
+      </div>
+      <div
+        className="topology-workload-details"
+        data-testid="deployment-details"
```
Comment (debsmita1) [inline plugins/topology/src/components/Topology/TopologySideBar/TopologyVirtualMachineDetails.tsx:65]:
```suggestion
        data-testid="virtualmachine-details"
```

File: plugins/topology/src/models.ts [hunk 8]
```diff
@@ -190,4 +214,6 @@ export const resourceModels = {
   [TaskRunModel.kind]: TaskRunModel,
   [CheClusterModel.kind]: CheClusterModel,
   [VirtualMachineModel.kind]: VirtualMachineModel,
+  [VirtualMachineInstanceModel.kind]: VirtualMachineInstanceModel,
+  [TemplateModel.kind]: TemplateModel,
```
Comment (debsmita1) [inline plugins/topology/src/models.ts:218]: If you are planning to use these resources, you'll need to update the Readme to include them as k8s custom resources.

File: plugins/topology/src/types/api/V1VirtualMachineVolumeRequest.ts [hunk 9]
```diff
@@ -0,0 +1,9 @@
+import { V1AddVolumeOptions } from './V1AddVolumeOptions';
+import { V1RemoveVolumeOptions } from './V1RemoveVolumeOptions';
+
+export interface V1VirtualMachineVolumeRequest {
```
Comment (debsmita1) [inline plugins/topology/src/types/api/V1VirtualMachineVolumeRequest.ts:4]: Where is this used ?

File: plugins/topology/src/types/api/V1Volume.ts [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+import { V1SecretVolumeSource } from './V1SecretVolumeSource';
+import { V1ServiceAccountVolumeSource } from './V1ServiceAccountVolumeSource';
+import { V1SysprepSource } from './v1SysprepSource';
+
+/**
+ * Volume represents a named volume in a vmi.
+ * @export
+ * @interface V1Volume
+ */
+export interface V1Volume {
```
Comment (debsmita1) [inline plugins/topology/src/types/api/V1Volume.ts:32]: Where is this used ?

**Review by debsmita1 — CHANGES_REQUESTED** [review]

**Comment by debsmita1** [comment]
The `Service` created by the VM doesn't show up on the sidepanel

you would need to add
```
case 'VirtualMachine':
        return resource?.spec?.template;
```
in the `getPodTemplate`  util

File: plugins/topology/src/components/Topology/TopologySideBar/TopologyVirtualMachineDetails.tsx (hunk 1 above)
Comment (its-mitesh-kumar) [reply plugins/topology/src/components/Topology/TopologySideBar/TopologyVirtualMachineDetails.tsx]: updated correct data

File: plugins/topology/src/components/Topology/TopologySideBar/TopologyVirtualMachineDetails.tsx (hunk 2 above)
Comment (its-mitesh-kumar) [reply plugins/topology/src/components/Topology/TopologySideBar/TopologyVirtualMachineDetails.tsx]: removed

File: plugins/topology/src/components/Topology/TopologySideBar/TopologyVirtualMachineDetails.tsx (hunk 3 above)
Comment (its-mitesh-kumar) [reply plugins/topology/src/components/Topology/TopologySideBar/TopologyVirtualMachineDetails.tsx]: removed

File: plugins/topology/src/components/Topology/TopologySideBar/TopologyVirtualMachineDetails.tsx (hunk 4 above)
Comment (its-mitesh-kumar) [reply plugins/topology/src/components/Topology/TopologySideBar/TopologyVirtualMachineDetails.tsx:124]: Done

File: plugins/topology/src/vm-models.ts (hunk 5 above)
Comment (its-mitesh-kumar) [reply plugins/topology/src/vm-models.ts:17]: We don't show RepilicaControllers

File: plugins/topology/src/vm-models.ts (hunk 6 above)
Comment (its-mitesh-kumar) [reply plugins/topology/src/vm-models.ts:23]: Done

File: plugins/topology/src/models.ts (hunk 8 above)
Comment (its-mitesh-kumar) [reply plugins/topology/src/models.ts:216]: We are not fetching template , but just we are showing the name , abbreviation and color .

**Review by debsmita1 — APPROVED** [review]
verified this locally
/lgtm

**Review by invincibleJai — APPROVED** [review]

**Review by jerolimov — APPROVED** [review]

**Pull request merged**